File: packages/sdk-node/src/core/rate-limiter.ts
/**
 * Token-bucket rate limiter — same algorithm across all 3 RateGuard SDKs.
 *
 * Algorithm: Token Bucket (RFC standards track, used by Kong, Envoy, AWS API Gateway)
 * - max_tokens = burst (bucket capacity)
 * - refill_rate = requests_per_second (tokens added per second)
 * - On each request: refill = elapsed × refill_rate, clamp to max_tokens
 * - Allow if tokens >= 1.0, consume 1 token
 * - Retry-after: ceil(deficit_seconds) rounded up to the next whole second,
 *   floored at 1000ms — matches the Go and Python SDKs exactly (cross-checked
 *   by conformance/token_bucket_vectors.json) so a client backing off on
 *   retry_after_ms behaves identically regardless of which SDK served it.
 *
 * Source: https://en.wikipedia.org/wiki/Token_bucket
 */

import { BoundedCache } from './bounded-cache.js';
import type { BucketState, Clock, RateLimitDecision, RateLimitOptions } from '../types.js';

interface Bucket {
  tokens: number; // float — fractional tokens for smooth refill
  last: number;   // ms timestamp
}

interface RemoteRateLimitResponse {
  allowed: boolean;
  remaining?: number;
  retry_after_ms?: number;
  retryAfterMs?: number;
}

/**
 * In-process token-bucket rate limiter with optional remote control-plane fallback.
 */
export class RateLimiter {
  private readonly clock: Clock;
  private readonly buckets: BoundedCache<string, Bucket>;

  constructor(options: { clock: Clock; capacity?: number }) {
    this.clock = options.clock;
    this.buckets = new BoundedCache<string, Bucket>(options.capacity ?? 50_000);
  }

  async allow(
    key: string,
    options: Required<RateLimitOptions> & { apiKey: string | undefined },
  ): Promise<RateLimitDecision> {
    const local = this.allowLocal(key, options);
    if (!local.allowed) {
      return local;
    }

    if (!options.remoteRateLimitEndpoint) {
      return local;
    }

    try {
      const response = await fetch(options.remoteRateLimitEndpoint, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          ...(options.apiKey ? { authorization: `Bearer ${options.apiKey}` } : {}),
        },
        body: JSON.stringify({
          key,
          requests_per_second: options.requestsPerSecond,
          burst: options.burst,
          window_ms: options.windowMs,
        }),
      });

      if (!response.ok) {
        return remoteUnavailable(local);
      }

      const body: unknown = await response.json();
      if (isRemoteRateLimitResponse(body)) {
        return {
          allowed: body.allowed,
          applied: true,
          remaining: typeof body.remaining === 'number' ? body.remaining : local.remaining,
          retryAfterMs:
            typeof body.retry_after_ms === 'number'
              ? body.retry_after_ms
              : typeof body.retryAfterMs === 'number'
                ? body.retryAfterMs
                : local.retryAfterMs,
          limit: local.limit,
          degraded: false,
        };
      }

      return remoteUnavailable(local);
    } catch {
      return remoteUnavailable(local);
    }
  }

  /**
   * Reports what allow() would decide right now WITHOUT consuming a token.
   * Pre-flight queries (MCP tools, dashboards) must use peek, never allow.
   */
  peek(key: string, options: Required<RateLimitOptions>): RateLimitDecision {
    const rps = options.requestsPerSecond;
    const burst = options.burst;
    if (rps <= 0 || burst <= 0) {
      return { allowed: true, applied: false, remaining: -1, retryAfterMs: 0, limit: -1, degraded: false };
    }

    const now = this.clock.now();
    const bucket = this.buckets.get(key);
    if (!bucket) {
      return { allowed: true, applied: true, remaining: burst, retryAfterMs: 0, limit: rps, degraded: false };
    }

    let tokens = bucket.tokens;
    if (now - bucket.last > 600_000) {
      tokens = burst;
    } else {
      const elapsed = (now - bucket.last) / 1000;
      if (elapsed > 0) {
        tokens = Math.min(burst, tokens + elapsed * rps);
      }
    }

    if (tokens < 1.0) {
      const deficit = (1.0 - tokens) / rps;
      return {
        allowed: false,
        applied: true,
        remaining: 0,
        retryAfterMs: Math.max(1000, Math.ceil(deficit) * 1000),
        limit: rps,
        degraded: false,
      };
    }

    return {
      allowed: true,
      applied: true,
      remaining: Math.max(0, Math.floor(tokens)),
      retryAfterMs: 0,
      limit: rps,
      degraded: false,
    };
  }

  private allowLocal(key: string, options: Required<RateLimitOptions>): RateLimitDecision {
    return this.incrementLocal(key, options, 1);
  }

  /**
   * Consumes n tokens atomically. allowLocal(key, options) is equivalent to
   * incrementLocal(key, options, 1). Used when a single call costs more than
   * one unit of the limit — e.g. an LLM request billed by estimated token
   * count rather than by call count.
   */
  private incrementLocal(key: string, options: Required<RateLimitOptions>, n: number): RateLimitDecision {
    const rps = options.requestsPerSecond;
    const burst = options.burst;
    if (rps <= 0 || burst <= 0) {
      return { allowed: true, applied: false, remaining: -1, retryAfterMs: 0, limit: -1, degraded: false };
    }

    const now = this.clock.now();

    let bucket = this.buckets.get(key);
    if (!bucket) {
      bucket = { tokens: burst, last: now };
      this.buckets.set(key, bucket);
    }

    // Idle bucket: reset after 10 minutes of inactivity
    if (now - bucket.last > 600_000) {
      bucket.tokens = burst;
      bucket.last = now;
    }

    // Token bucket refill: tokens = min(burst, tokens + elapsed × rps)
    const elapsed = (now - bucket.last) / 1000; // ms → seconds
    if (elapsed > 0) {
      bucket.tokens = Math.min(burst, bucket.tokens + elapsed * rps);
      bucket.last = now;
    }

    // Deny if not enough tokens
    if (bucket.tokens < n) {
      const deficit = (n - bucket.tokens) / rps;
      const retryAfterMs = Math.max(1000, Math.ceil(deficit) * 1000);
      return {
        allowed: false,
        applied: true,
        remaining: 0,
        retryAfterMs,
        limit: rps,
        degraded: false,
      };
    }

    // Allow: consume n tokens
    bucket.tokens -= n;
    return {
      allowed: true,
      applied: true,
      remaining: Math.max(0, Math.floor(bucket.tokens)),
      retryAfterMs: 0,
      limit: rps,
      degraded: false,
    };
  }

  /**
   * Consumes n tokens atomically, bypassing the optional remote control-plane
   * fallback that allow() performs. increment(key, options, 1) behaves
   * identically to the local decision allow() would make.
   */
  increment(key: string, options: Required<RateLimitOptions>, n: number): RateLimitDecision {
    return this.incrementLocal(key, options, n);
  }

  /**
   * Returns the current bucket state for key without consuming anything.
   * Never creates bucket state for unseen keys.
   */
  get(key: string, options: Required<RateLimitOptions>): BucketState {
    const rps = options.requestsPerSecond;
    const burst = options.burst;
    const bucket = this.buckets.get(key);
    if (!bucket) {
      return { tokens: burst, capacity: burst, limit: rps };
    }

    const now = this.clock.now();
    let tokens = bucket.tokens;
    if (now - bucket.last > 600_000) {
      tokens = burst;
    } else {
      const elapsed = (now - bucket.last) / 1000;
      if (elapsed > 0) {
        tokens = Math.min(burst, tokens + elapsed * rps);
      }
    }

    return { tokens, capacity: burst, limit: rps };
  }

  /**
   * Clears key's bucket; the next access starts from a full bucket.
   */
  reset(key: string): void {
    this.buckets.delete(key);
  }
}

function remoteUnavailable(local: RateLimitDecision): RateLimitDecision {
  return {
    ...local,
    allowed: false,
    applied: false,
    remaining: 0,
    retryAfterMs: 0,
    degraded: true,
  };
}

function isRemoteRateLimitResponse(value: unknown): value is RemoteRateLimitResponse {
  if (!value || typeof value !== 'object') {
    return false;
  }
  const record = value as Record<string, unknown>;
  return typeof record.allowed === 'boolean';
}

File: packages/sdk-python/rateguard/core/rate_limiter.py
"""
Token-bucket rate limiter — same algorithm across all 3 RateGuard SDKs.

Algorithm: Token Bucket (RFC standards track, used by Kong, Envoy, AWS API Gateway)
- max_tokens = burst (bucket capacity)
- refill_rate = requests_per_second (tokens added per second)
- On each request: refill = elapsed_seconds × refill_rate, clamp to max_tokens
- Allow if tokens >= 1.0, consume 1 token
- Retry-after: time until bucket refills to 1.0 tokens

Source: https://en.wikipedia.org/wiki/Token_bucket
"""

from __future__ import annotations

from dataclasses import dataclass
from math import ceil
from threading import RLock
from typing import TYPE_CHECKING

from ..types import BucketState, Clock, RateLimitDecision, RateLimitOptions
from .bounded_cache import BoundedCache

if TYPE_CHECKING:
    pass


@dataclass(slots=True)
class _Bucket:
    tokens: float
    last: float  # ms timestamp


class RateLimiter:
    """In-process token-bucket rate limiter with bounded cache.

    All 3 RateGuard SDKs (Go, Node, Python) use the same token-bucket
    algorithm for deterministic, predictable behavior across languages.
    """

    def __init__(self, clock: Clock, capacity: int = 50_000) -> None:
        self._clock = clock
        self._buckets = BoundedCache[str, _Bucket](capacity)
        self._lock = RLock()

    def allow(self, key: str, options: RateLimitOptions, **kwargs: object) -> RateLimitDecision:
        rps = options.requests_per_second or 0
        burst = options.burst or 0
        return self._increment_token_bucket(key, rps, burst, 1.0)

    async def allow_async(self, key: str, options: RateLimitOptions, **kwargs: object) -> RateLimitDecision:
        return self.allow(key, options)

    def increment(self, key: str, options: RateLimitOptions, n: float) -> RateLimitDecision:
        """Consume n tokens atomically. increment(key, options, 1) behaves
        identically to allow(key, options). Used when a single call costs
        more than one unit of the limit — e.g. an LLM request billed by
        estimated token count rather than by call count.
        """
        rps = options.requests_per_second or 0
        burst = options.burst or 0
        return self._increment_token_bucket(key, rps, burst, n)

    def get(self, key: str, options: RateLimitOptions) -> BucketState:
        """Return the current bucket state for key without consuming
        anything. Never creates bucket state for unseen keys."""
        rps = options.requests_per_second or 0
        burst = options.burst or 0

        with self._lock:
            bucket = self._buckets.get(key)

        if bucket is None:
            return BucketState(tokens=float(burst), capacity=burst, limit=rps)

        now = self._clock.now()
        tokens = bucket.tokens
        if now - bucket.last > 600_000:
            tokens = float(burst)
        else:
            elapsed = (now - bucket.last) / 1000.0
            if elapsed > 0:
                tokens = min(float(burst), tokens + elapsed * float(rps))

        return BucketState(tokens=tokens, capacity=burst, limit=rps)

    def reset(self, key: str) -> None:
        """Clear key's bucket; the next access starts from a full bucket."""
        with self._lock:
            self._buckets.delete(key)

    def peek(self, key: str, options: RateLimitOptions) -> RateLimitDecision:
        """Report what allow() would decide right now WITHOUT consuming a token.

        Pre-flight queries (MCP tools, dashboards) must use peek, never allow.
        Never creates bucket state for unseen keys.
        """
        rps = options.requests_per_second or 0
        burst = options.burst or 0
        if rps <= 0 or burst <= 0:
            return RateLimitDecision(True, False, -1, 0, -1, False)

        now = self._clock.now()  # ms

        with self._lock:
            bucket = self._buckets.get(key)

        if bucket is None:
            return RateLimitDecision(True, True, burst, 0, rps, False)

        tokens = bucket.tokens
        if now - bucket.last > 600_000:
            tokens = float(burst)
        else:
            elapsed = (now - bucket.last) / 1000.0
            if elapsed > 0:
                tokens = min(float(burst), tokens + elapsed * float(rps))

        if tokens < 1.0:
            deficit = 1.0 - tokens
            retry_ms = max(1000, int(ceil(deficit / float(rps)) * 1000))
            return RateLimitDecision(False, True, 0, retry_ms, rps, False)

        return RateLimitDecision(True, True, max(0, int(tokens)), 0, rps, False)

    def _increment_token_bucket(
        self, key: str, rps: int, burst: int, n: float
    ) -> RateLimitDecision:
        """Consume n tokens atomically under the token bucket.

        Formula: tokens = min(burst, tokens + elapsed × rps)
        Allow: tokens >= n → consume n
        Deny: retry_after = ceil((n - tokens) / rps) × 1000 ms
        """
        if rps <= 0 or burst <= 0:
            return RateLimitDecision(True, False, -1, 0, -1, False)

        now = self._clock.now()  # ms

        with self._lock:
            bucket = self._buckets.get_or_create(
                key, lambda: _Bucket(tokens=float(burst), last=now)
            )

        # Idle bucket: reset after 10 minutes of inactivity
        if now - bucket.last > 600_000:  # 10 min in ms
            bucket.tokens = float(burst)
            bucket.last = now

        # Token bucket refill
        elapsed = (now - bucket.last) / 1000.0  # ms → seconds
        if elapsed > 0:
            bucket.tokens = min(float(burst), bucket.tokens + elapsed * float(rps))
            bucket.last = now

        # Deny if not enough tokens
        if bucket.tokens < n:
            deficit = n - bucket.tokens
            retry_sec = ceil(deficit / float(rps))
            retry_ms = max(1000, int(retry_sec * 1000))
            return RateLimitDecision(
                allowed=False,
                applied=True,
                remaining=0,
                retry_after_ms=retry_ms,
                limit=rps,
                degraded=False,
            )

        # Allow: consume n tokens
        bucket.tokens -= n
        return RateLimitDecision(
            allowed=True,
            applied=True,
            remaining=max(0, int(bucket.tokens)),
            retry_after_ms=0,
            limit=rps,
            degraded=False,
        )
